src/validation.cpp

```diff
@@ -340,21 +340,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
     // the disconnectpool that were added back and cleans up the mempool state.
     m_mempool->UpdateTransactionsFromBlock(vHashUpdate);
 
-    const auto check_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it)
+    // Check whether the transaction is still final and, if it spends a coinbase output, mature.
+    // Update the entry's cached LockPoints if needed. If this returns true, the tx would be invalid
+    // in the next block, so the mempool must remove this entry and all of its descendants.
+    const auto filter_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it)
         EXCLUSIVE_LOCKS_REQUIRED(m_mempool->cs, ::cs_main) {
-        bool should_remove = false;
         AssertLockHeld(m_mempool->cs);
         AssertLockHeld(::cs_main);
         const CTransaction& tx = it->GetTx();
         LockPoints lp = it->GetLockPoints();
         const bool validLP{TestLockPointValidity(m_chain, lp)};
         CCoinsViewMemPool view_mempool(&CoinsTip(), *m_mempool);
-        if (!CheckFinalTx(m_chain.Tip(), tx, flags)
-            || !CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
-            // Note if CheckSequenceLocks fails the LockPoints may still be invalid
-            // So it's critical that we remove the tx and not depend on the LockPoints.
-            should_remove = true;
-        } else if (it->GetSpendsCoinbase()) {
+        // The transaction must be final.
+        if (!CheckFinalTx(m_chain.Tip(), tx, flags)) return true;
+        // Note if CheckSequenceLocks fails the LockPoints may still be invalid
+        // So it's critical that we remove the tx and not depend on the LockPoints.
+        if (!CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
+            return true;
+        } else if (!validLP) {
```

Comment:
```suggestion
        }  if (!validLP) {
```

nit: no need for else. I find it confusing in early return statements.

Reply:
See rationale here: https://github.com/bitcoin/bitcoin/pull/23683#discussion_r772879178 , but you're right that they've equivalent, so both are fine.

Reply:
Why this branch--`if (!validLP)`--has been moved before `if (it->GetSpendsCoinbase())` ?

Reply:
Because logically it's connected to the previous if statement - if `CheckSequenceLocks()` returns false then this filter function fails immediately. If it returns true, then the lockpoints may be invalid so they should be updated.

Functionally this is equivalent, but it groups the related functionality together.

Reply:
Clarified the comments so it should be more clear to a first-time reader why the branches are linked